File: av-bridge-cloud/internal/portalauth/permissions.go
package portalauth

// Permission catalogue — the canonical list of capability keys the tenant
// RBAC engine understands. Every value in role_permissions.permission must
// be a member of KnownPermissions or the row is unreachable (the middleware
// will silently ignore an unknown key, which is a bug in the row not in
// the middleware).
//
// The migration 0016_rbac_roles.sql seeds the three system-default roles
// using these exact strings — keep both in sync. Adding a new permission is
// a two-step change:
//
//   1. Add the constant + register in KnownPermissions here.
//   2. Update the seeds in db/customers.seedSystemRoles() and in the
//      relevant migration so existing customers get the new permission on
//      the appropriate role (usually admin).
//
// The portal has a mirror catalogue in src/lib/permissions.ts with UI
// metadata (category label, human description). When adding permissions
// here, add the matching entry there so the /roles page renders them.

// Read permissions — a user with these can see the corresponding data.
const (
	PermViewDashboard     = "view.dashboard"
	PermViewAudit         = "view.audit"
	PermViewReports       = "view.reports"
	PermViewFirmware      = "view.firmware"
	PermViewNotifications = "view.notifications"
	PermViewUsers         = "view.users"
	PermViewAssets        = "view.assets"
)

// Device control permissions — send commands, reconnect, fan-out.
const (
	PermCommandDevice    = "command.device"
	PermCommandBulk      = "command.bulk"
	PermReconnectDevice  = "reconnect.device"
)

// Alert lifecycle permissions — an ops-team caller usually has both.
const (
	PermAlertAcknowledge = "alert.acknowledge"
	PermAlertResolve     = "alert.resolve"
)

// Fleet + hierarchy management.
const (
	PermDeviceCRUD    = "device.crud"
	PermHierarchyCRUD = "hierarchy.crud"
	PermAssetCRUD     = "asset.crud"
	// collector.crud — pre-provision a new bridge (collector) from the
	// portal and mint an enrollment token an engineer runs on-site.
	// Kept distinct from device.crud so a customer admin can be
	// trusted with device inventory without also being trusted to
	// stand up new bridges (which is more sensitive: the bridge is a
	// long-lived HMAC-authenticated identity).
	PermCollectorCRUD = "collector.crud"
)

// Outbound notifications — channel CRUD is the manage side; test is the
// send-a-test-message operator side, kept separate so an on-call operator
// can verify their own channel without full admin rights.
const (
	PermNotificationCRUD = "notification.crud"
	PermNotificationTest = "notification.test"
)

// Firmware policy — per-(make, model) target_version + docs_url.
const PermFirmwareTargetCRUD = "firmware_target.crud"

// User + role management. Delete is separated from update so an operator-
// like role can be defined with the ability to disable/enable users but
// not permanently delete them (via update.disabled=true without owning
// user.delete).
const (
	PermUserCreate        = "user.create"
	PermUserUpdate        = "user.update"
	PermUserResetPassword = "user.reset_password"
	PermUserDelete        = "user.delete"
	PermRoleCRUD          = "role.crud"
	// role_mapping.manage — CRUD over the Entra group → role mappings
	// consulted at JIT time. Kept distinct from role.crud so a tenant
	// can grant the "who defines groups → roles" capability separately
	// from "who defines what roles mean", though in practice the
	// system-default admin role holds both.
	PermRoleMappingManage = "role_mapping.manage"
)

// Tenant-branding management. Reads are open to any authenticated user in
// the tenant (the whole portal renders in that customer's colours, so
// gating the read serves no purpose); writes gate on this permission.
const PermBrandingUpdate = "branding.update"

// Nightly lifecycle (Room Readiness). view lets a user see schedules,
// routines, and run history; manage lets an admin edit schedules, author
// routines, set per-room overrides. Operators + viewers get view only.
const (
	PermNightlyView   = "nightly.view"
	PermNightlyManage = "nightly.manage"
)

// Public API tokens — programmatic access. view lets a role see the
// token list + last-used timestamps (useful for compliance / audit
// dashboards); manage lets an admin mint, name, and revoke tokens.
// Both are admin-only by seed — an integrating system's key is a
// long-lived tenant credential, not a per-user preference.
const (
	PermAPITokenView   = "api_token.view"
	PermAPITokenManage = "api_token.manage"
)

// KnownPermissions is the closed set of valid permission keys — used by the
// roles CRUD handler to reject unknown strings, and by the permission
// engine to fail-closed if a role somehow references a permission that has
// been removed from the code without a matching data migration.
var KnownPermissions = map[string]struct{}{
	PermViewDashboard:      {},
	PermViewAudit:          {},
	PermViewReports:        {},
	PermViewFirmware:       {},
	PermViewNotifications:  {},
	PermViewUsers:          {},
	PermCommandDevice:      {},
	PermCommandBulk:        {},
	PermReconnectDevice:    {},
	PermAlertAcknowledge:   {},
	PermAlertResolve:       {},
	PermDeviceCRUD:         {},
	PermHierarchyCRUD:      {},
	PermCollectorCRUD:      {},
	PermNotificationCRUD:   {},
	PermNotificationTest:   {},
	PermFirmwareTargetCRUD: {},
	PermUserCreate:         {},
	PermUserUpdate:         {},
	PermUserResetPassword:  {},
	PermUserDelete:         {},
	PermRoleCRUD:           {},
	PermRoleMappingManage:  {},
	PermBrandingUpdate:     {},
	PermViewAssets:         {},
	PermAssetCRUD:          {},
	PermNightlyView:        {},
	PermNightlyManage:      {},
	PermAPITokenView:       {},
	PermAPITokenManage:     {},
}

// IsKnownPermission returns true if the string is a member of the
// catalogue above. Used by role CRUD to validate incoming lists.
func IsKnownPermission(p string) bool {
	_, ok := KnownPermissions[p]
	return ok
}

// VendorRolePermissions holds the fixed permission set for each vendor
// role. Vendors have historically been a single "admin bypasses
// everything" tier; this map introduces role-based restrictions so a
// helpdesk operator or viewer sees a limited surface even though they
// still act cross-tenant via X-Customer-Scope.
//
// Structure:
//   - "admin": nil — signals "bypass all checks", identical to the
//     pre-role-aware behaviour so existing seed data keeps working.
//   - "operator": read + operational writes (commands, alert lifecycle,
//     notification test-sends). No structural CRUD, no user or role
//     management, no branding, no magic links.
//   - "viewer": read-only. All view.* + nightly.view. No writes.
//
// A vendor row with a role not in this map (e.g. a stale value from an
// old install) falls back to viewer — fail-closed rather than fall
// through to admin.
var VendorRolePermissions = map[string]map[string]struct{}{
	// nil = bypass (see HasPermission). Represented as a distinct nil
	// entry so the presence check tells apart "unknown role" from
	// "known admin".
	"admin": nil,
	"operator": {
		PermViewDashboard:     {},
		PermViewAudit:         {},
		PermViewReports:       {},
		PermViewFirmware:      {},
		PermViewNotifications: {},
		PermViewUsers:         {},
		PermViewAssets:        {},
		PermCommandDevice:     {},
		PermCommandBulk:       {},
		PermReconnectDevice:   {},
		PermAlertAcknowledge:  {},
		PermAlertResolve:      {},
		PermNotificationTest:  {},
		PermNightlyView:       {},
	},
	"viewer": {
		PermViewDashboard:     {},
		PermViewAudit:         {},
		PermViewReports:       {},
		PermViewFirmware:      {},
		PermViewNotifications: {},
		PermViewUsers:         {},
		PermViewAssets:        {},
		PermNightlyView:       {},
	},
}

// VendorRoleBypasses reports whether a vendor role should short-circuit
// every permission check to allowed. Only "admin" does. Kept as a
// helper so callers don't have to open-code the nil check.
func VendorRoleBypasses(role string) bool {
	set, ok := VendorRolePermissions[role]
	return ok && set == nil
}

// VendorRoleHasPermission checks whether a vendor role's fixed
// permission set includes perm. "admin" always returns true (bypass);
// unknown roles fall back to the viewer set for fail-closed safety.
func VendorRoleHasPermission(role, perm string) bool {
	set, ok := VendorRolePermissions[role]
	if !ok {
		set = VendorRolePermissions["viewer"]
	}
	if set == nil { // admin bypass
		return true
	}
	_, ok = set[perm]
	return ok
}

File: av-bridge-portal/src/lib/permissions.ts
// Portal-side mirror of av-bridge-cloud/internal/portalauth/permissions.go.
//
// The Go catalogue is the source of truth — the backend validates every
// incoming permission key against IsKnownPermission and rejects unknowns
// with a 400. This file adds UI metadata (category label, human title,
// description) so the /roles page permission matrix renders sensibly.
//
// When adding a permission on the backend, ALSO add it here so the UI can
// display + toggle it. When removing one, remove here too; a permission
// row without a matching entry falls out of the matrix silently (the UI
// only renders keys it recognises).

export type PermissionCategory =
  | "Reads"
  | "Device control"
  | "Alerts"
  | "Fleet management"
  | "Notifications"
  | "Firmware"
  | "Users & Roles"
  | "Tenant"
  | "Assets"
  | "Room Readiness"
  | "Public API";

export interface PermissionDef {
  key: string;
  title: string;
  description: string;
  category: PermissionCategory;
}

// Ordering here drives display order in the /roles matrix. Categories are
// listed in the order they first appear.
export const PERMISSIONS: readonly PermissionDef[] = [
  // Reads
  { key: "view.dashboard",     title: "View dashboard",       description: "Fleet summary, device list, telemetry, live event stream.", category: "Reads" },
  { key: "view.audit",         title: "View audit log",       description: "See who did what and when across the tenant.",             category: "Reads" },
  { key: "view.reports",       title: "View reports",         description: "Device uptime and room activity reports, CSV export.",     category: "Reads" },
  { key: "view.firmware",      title: "View firmware",        description: "Fleet firmware histogram + target versions.",              category: "Reads" },
  { key: "view.notifications", title: "View notifications",   description: "List the notification channels configured for the tenant.", category: "Reads" },
  { key: "view.users",         title: "View users",           description: "List of users in the tenant and their assigned roles.",   category: "Reads" },
  { key: "view.assets",        title: "View assets",          description: "List physical assets (monitored + not) tracked in the tenant's rooms.", category: "Reads" },

  // Device control
  { key: "command.device",    title: "Send device commands", description: "Issue single-device commands (power, input, mute, dial...).", category: "Device control" },
  { key: "command.bulk",      title: "Bulk command fan-out", description: "Send the same command to many devices at once (up to 200).",  category: "Device control" },
  { key: "reconnect.device",  title: "Reconnect device",     description: "Force the bridge to drop and re-open a device connection.",   category: "Device control" },

  // Alerts
  { key: "alert.acknowledge", title: "Acknowledge alerts",   description: "Move an open alert to acknowledged so on-call knows it's being handled.", category: "Alerts" },
  { key: "alert.resolve",     title: "Resolve alerts",       description: "Close an alert as resolved. Records who resolved it.",                    category: "Alerts" },

  // Fleet management
  { key: "device.crud",       title: "Manage devices",       description: "Add, edit, and remove devices from the tenant.",           category: "Fleet management" },
  { key: "hierarchy.crud",    title: "Manage locations",     description: "Regions, locations, buildings, rooms — full CRUD.",         category: "Fleet management" },
  { key: "collector.crud",    title: "Manage collectors",    description: "Pre-provision on-prem bridges from the portal and issue enrollment tokens for site setup.", category: "Fleet management" },

  // Notifications
  { key: "notification.crud", title: "Manage notification channels", description: "Add, edit, and remove email / Teams / webhook channels.",       category: "Notifications" },
  { key: "notification.test", title: "Test notification channel",     description: "Send a test message to a channel to verify it's configured right.", category: "Notifications" },

  // Firmware
  { key: "firmware_target.crud", title: "Manage firmware targets", description: "Set the target firmware version + docs URL per (make, model).", category: "Firmware" },

  // Users & Roles
  { key: "user.create",         title: "Create users",         description: "Add new users to the tenant.",                                     category: "Users & Roles" },
  { key: "user.update",         title: "Update users",         description: "Change a user's full name, role assignments, or enabled state.",   category: "Users & Roles" },
  { key: "user.reset_password", title: "Reset user passwords", description: "Set a new password for a user without knowing the current one.",   category: "Users & Roles" },
  { key: "user.delete",         title: "Delete users",         description: "Permanently remove a user from the tenant.",                       category: "Users & Roles" },
  { key: "role.crud",           title: "Manage roles",         description: "Create, edit, and delete custom roles inside this tenant.",        category: "Users & Roles" },
  { key: "role_mapping.manage", title: "Manage sign-in mappings", description: "Map Entra security groups to roles so SSO users land with the right permissions on first sign-in.", category: "Users & Roles" },
  // Tenant
  { key: "branding.update",     title: "Update branding",      description: "Upload the tenant logo, change the accent colour and portal display name.", category: "Tenant" },
  // Assets
  { key: "asset.crud",          title: "Manage assets",        description: "Add, edit, and remove assets (physical inventory) in the tenant.", category: "Assets" },
  // Room Readiness (nightly lifecycle)
  { key: "nightly.view",        title: "View Room Readiness",  description: "See nightly schedules, routines, and run history.",                       category: "Room Readiness" },
  { key: "nightly.manage",      title: "Manage Room Readiness", description: "Edit the nightly schedule, room overrides, and test routines.",         category: "Room Readiness" },
  // Public API
  { key: "api_token.view",      title: "View API tokens",       description: "See the list of API tokens minted for programmatic access, and their last-used timestamps.", category: "Public API" },
  { key: "api_token.manage",    title: "Manage API tokens",     description: "Mint, name, and revoke API tokens used by external systems to read from the tenant.",       category: "Public API" },
];

// PERMISSIONS_BY_CATEGORY is what the matrix renders — same items, grouped
// for the sectioned checkbox grid.
export const PERMISSIONS_BY_CATEGORY: { category: PermissionCategory; items: readonly PermissionDef[] }[] = (() => {
  const seen = new Set<PermissionCategory>();
  const order: PermissionCategory[] = [];
  for (const p of PERMISSIONS) {
    if (!seen.has(p.category)) {
      seen.add(p.category);
      order.push(p.category);
    }
  }
  return order.map((category) => ({
    category,
    items: PERMISSIONS.filter((p) => p.category === category),
  }));
})();

// KNOWN_PERMISSION_KEYS is the exact string set the backend accepts — used
// to filter out unknown keys arriving in a role's permissions[] payload.
export const KNOWN_PERMISSION_KEYS = new Set(PERMISSIONS.map((p) => p.key));
